### ISSUE TITLE
Compatibilty issue with laravel postgresql enhanced

### ISSUE THREAD

**Issue by Sergiobop** [staudenmeir/laravel-cte#41]
There are issues with https://github.com/tpetry/laravel-postgresql-enhanced:

`
Tpetry\PostgresqlEnhanced\Query\Grammar::prependCtes(): Argument #1 ($query) must be of type Tpetry\PostgresqlEnhanced\Query\Builder, Staudenmeir\LaravelCte\Query\Builder given, called in vendor/tpetry/laravel-postgresql-enhanced/src/Query/Grammar.php on...: Tpetry\\PostgresqlEnhanced\\Query\\Grammar::prependCtes(): Argument #1 ($query) must be of type Tpetry\\PostgresqlEnhanced\\Query\\Builder, Staudenmeir\\LaravelCte\\Query\\Builder given, called in vendor/tpetry/laravel-postgresql-enhanced/src/Query/Grammar.php on line 54 at vendor/tpetry/laravel-postgresql-enhanced/src/Query/GrammarCte.php:36)
`

**Comment by staudenmeir** [comment]
Hi @Sergiobop,
`tpetry/laravel-postgresql-enhanced` added support for common table expressions a while ago: https://github.com/tpetry/laravel-postgresql-enhanced#common-table-expressions-cte

It's probably best to remove my package and only use the other one. It should offer the same features.

**Comment by Sergiobop** [comment]
Hi @staudenmeir , first of all, thank you for your work and your fast response!

Unfortunately, i can't uninstall this package, since i'm using [laravel-adjacency-list](https://github.com/staudenmeir/laravel-adjacency-list) and this one is a dependency

**Comment by staudenmeir** [comment]
Good point. What query are you executing when you get this error?

**Comment by Sergiobop** [comment]
I have 2 models: Category and Thing.

Thing belongsToMany categories.

Category uses the trait HasRecursiveRelationships.

I'm doing something like this, filtering the things which has one given category:

```
Thing::query()
     ->whereHas('categories', function (Builder $query) use ($request) {
                  $query->where('id', $request->input('category_id'));
      })
     ->get()
 ```

**Comment by staudenmeir** [comment]
Please share the whole error stacktrace. Did you only install `tpetry/laravel-postgresql-enhanced` or did also add any of its traits to your models (if the package has any)?

**Comment by staudenmeir** [comment]
I couldn't reproduce the issue in a quick test, but you can try the following workaround:
- Replace the `extra` attribute in your `composer.json` with this:
```json
    "extra": {
        "laravel": {
            "dont-discover": ["staudenmeir/laravel-cte"]
        }
    },
```
- Run `composer dump`.

**Comment by Sergiobop** [comment]
Full test;

I disabled postgresql-enhanced. Everything works, the query executed is something like this:

`
SELECT "things"."id", "things"."name", "things"."description" FROM "things" WHERE "things"."account_id" = '90913d87-50c3-49c5-a846-8cfe6373ad5b' and EXISTS (SELECT * FROM "categories" INNER JOIN "thing_category" ON "categories"."id" = "thing_category"."category_id" WHERE "things"."id" = "thing_category"."thing_id" and "id" = '97c45fd7-351d-41e0-ba71-eb3deadc5735') ORDER BY "things"."name" ASC LIMIT 25 OFFSET 0
`

Enable postgresql-enhanced again. Error:

```
 Tpetry\PostgresqlEnhanced\Query\Grammar::prependCtes(): Argument #1 ($query) must be of type Tpetry\PostgresqlEnhanced\Query\Builder, Staudenmeir\LaravelCte\Query\Builder given, called in /vendor/tpetry/laravel-postgresql-enhanced/src/Query/Grammar.php on line 54 {"userId":"1d7e27f6-70c9-4a74-bfff-eff0f685444e","exception":"[object] (TypeError(code: 0): Tpetry\\PostgresqlEnhanced\\Query\\Grammar::prependCtes(): Argument #1 ($query) must be of type Tpetry\\PostgresqlEnhanced\\Query\\Builder, Staudenmeir\\LaravelCte\\Query\\Builder given, called in /vendor/tpetry/laravel-postgresql-enhanced/src/Query/Grammar.php on line 54 at /vendor/tpetry/laravel-postgresql-enhanced/src/Query/GrammarCte.php:36)
[stacktrace]

#0 /vendor/tpetry/laravel-postgresql-enhanced/src/Query/Grammar.php(54): Tpetry\\PostgresqlEnhanced\\Query\\Grammar->prependCtes(Object(Staudenmeir\\LaravelCte\\Query\\Builder), 'select * from \"...')

#1 /vendor/laravel/framework/src/Illuminate/Database/Query/Grammars/Grammar.php(534): Tpetry\\PostgresqlEnhanced\\Query\\Grammar->compileSelect(Object(Staudenmeir\\LaravelCte\\Query\\Builder))

#2 /vendor/laravel/framework/src/Illuminate/Database/Query/Grammars/Grammar.php(221): Illuminate\\Database\\Query\\Grammars\\Grammar->whereExists(Object(Tpetry\\PostgresqlEnhanced\\Query\\Builder), Array)

#3 [internal function]: Illuminate\\Database\\Query\\Grammars\\Grammar->Illuminate\\Database\\Query\\Grammars\\{closure}(Array, 1)

#4 /vendor/laravel/framework/src/Illuminate/Collections/Arr.php(560): array_map(Object(Closure), Array, Array)

#5 /vendor/laravel/framework/src/Illuminate/Collections/Collection.php(719): Illuminate\\Support\\Arr::map(Array, Object(Closure))

#6 /vendor/laravel/framework/src/Illuminate/Database/Query/Grammars/Grammar.php(222): Illuminate\\Support\\Collection->map(Object(Closure))

#7 /vendor/laravel/framework/src/Illuminate/Database/Query/Grammars/Grammar.php(205): Illuminate\\Database\\Query\\Grammars\\Grammar->compileWheresToArray(Object(Tpetry\\PostgresqlEnhanced\\Query\\Builder))

#8 /vendor/laravel/framework/src/Illuminate/Database/Query/Grammars/Grammar.php(100): Illuminate\\Database\\Query\\Grammars\\Grammar->compileWheres(Object(Tpetry\\PostgresqlEnhanced\\Query\\Builder), Array)

#9 /vendor/laravel/framework/src/Illuminate/Database/Query/Grammars/Grammar.php(74): Illuminate\\Database\\Query\\Grammars\\Grammar->compileComponents(Object(Tpetry\\PostgresqlEnhanced\\Query\\Builder))

#10 /vendor/tpetry/laravel-postgresql-enhanced/src/Query/Grammar.php(54): Illuminate\\Database\\Query\\Grammars\\Grammar->compileSelect(Object(Tpetry\\PostgresqlEnhanced\\Query\\Builder))

#11 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2548): Tpetry\\PostgresqlEnhanced\\Query\\Grammar->compileSelect(Object(Tpetry\\PostgresqlEnhanced\\Query\\Builder))

#12 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2650): Illuminate\\Database\\Query\\Builder->toSql()
#13 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2638): Illuminate\\Database\\Query\\Builder->runSelect()

#14 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(3174): Illuminate\\Database\\Query\\Builder->Illuminate\\Database\\Query\\{closure}()

#15 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2639): Illuminate\\Database\\Query\\Builder->onceWithColumns(Array, Object(Closure))

#16 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2788): Illuminate\\Database\\Query\\Builder->get()

#17 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2747): Illuminate\\Database\\Query\\Builder->runPaginationCountQuery(Array)

#18 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(883): Illuminate\\Database\\Query\\Builder->getCountForPagination()
...
```


I disabled `staudenmeir/laravel-cte` as you said, adding it into the array of 'dont-discover'. Same error

**Comment by staudenmeir** [comment]
What version of `tpetry/laravel-postgresql-enhanced` are you using?

I can only reproduce the issue with `< 0.20.1` (https://github.com/tpetry/laravel-postgresql-enhanced/issues/38).

> I'm doing something like this, filtering the things which has one given category:

What's the actual query?

Here's a test repository where the query works: https://github.com/staudenmeir/laravel-cte-conflict-test

**Comment by Sergiobop** [comment]
Hi @staudenmeir , i was using 0.19. Let me try again with an updated version and i will let you know the results.

Thanks!

**Comment by staudenmeir** [comment]
Did it work?
